x-pack/plugins/lens/public/datatable_visualization_plugin/visualization.tsx

```diff
@@ -0,0 +1,215 @@
+  EuiFlexGroup,
+  EuiFlexItem,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { I18nProvider } from '@kbn/i18n/react';
+import {
+  SuggestionRequest,
+  Visualization,
+  VisualizationProps,
+  VisualizationSuggestion,
+} from '../types';
+import { NativeRenderer } from '../native_renderer';
+
+export interface DatatableVisualizationState {
+  columns: Array<{
+    id: string;
+    label: string;
+  }>;
+}
+
+export function DatatableConfigPanel(props: VisualizationProps<DatatableVisualizationState>) {
+  const { state, datasource, setState } = props;
+
+  return (
+    <div>
+      <EuiForm className="lnsConfigPanel">
+        {state.columns.map(({ id, label }, index) => {
+          const operation = datasource.getOperationForColumnId(id);
+          return (
+            <>
+              <EuiFormRow
+                key={id}
+                label={i18n.translate('xpack.lens.datatable.columnLabel', {
+                  defaultMessage: 'Column',
+                })}
+              >
+                <EuiFieldText
+                  value={label || ''}
+                  onChange={e => {
+                    const newColumns = [...state.columns];
+                    newColumns[index] = { ...newColumns[index], label: e.target.value };
+                    setState({
+                      ...state,
+                      columns: newColumns,
+                    });
+                  }}
+                  placeholder={
+                    operation
+                      ? operation.label
+                      : i18n.translate('xpack.lens.datatable.columnTitlePlaceholder', {
+                          defaultMessage: 'Title',
+                        })
+                  }
+                  aria-label={i18n.translate('xpack.lens.datatable.columnTitlePlaceholder', {
+                    defaultMessage: 'Title',
+                  })}
+                />
+              </EuiFormRow>
+
+              <EuiFormRow>
+                <EuiFlexGroup>
+                  <EuiFlexItem grow={true}>
+                    <NativeRenderer
+                      data-test-subj="lnsDatatable_xDimensionPanel"
+                      render={datasource.renderDimensionPanel}
+                      nativeProps={{
+                        columnId: id,
+                        dragDropContext: props.dragDropContext,
+                        filterOperations: () => true,
+                      }}
+                    />
+                  </EuiFlexItem>
+
+                  <EuiFlexItem grow={false}>
+                    <EuiButtonIcon
+                      size="s"
+                      color="warning"
+                      data-test-subj={`lns_datasourceDimensionPanel_remove_${id}`}
+                      iconType="trash"
+                      onClick={() => {
+                        datasource.removeColumnInTableSpec(id);
+                        const newColumns = [...state.columns];
+                        newColumns.splice(index);
```

Comment:
If you have three columns and press delete on the second one, it deletes both the second and the third because this is the default behavior of splice with one argument: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice Should be `newColumns.splice(index, 1);`

There are probably other places which have the same problem

Reply:
Will add a test for this